### DISCUSSION TITLE
A more complete sdn service

### DISCUSSION THREAD

**Discussion by PengfeiM** [coreemu/core#599] [Show and tell]
I have dumped the old sdn service which doesn't work so well and write  my own sdn service, which is improved and supports more ovs-vswitch  in emulation. 
I have completed the ovs part, I wonder if I can merge my work into CORE. 

Furthermore, I still have some problems that may need your advice. I will add them to this dicussion later.

**Comment by PatrickTague** [comment]
@PengfeiM curious to know, when you use the sdn service this way (which is much much cleaner than the original way), how do you point the SDN controller to interface with the OVS switches correctly?  for example, if i run the ryu-manager with one of the simple_switch applications, it adds the first rule but then ignores all packet_out messages from the controller.  do you happen to know why this happens or how to fix it?

**Reply by PengfeiM** [reply]
yes, we have fixed this problem. It's done by another member of our team. 
The general idea is as following: 
* a virtual eth of a node that added to the ovs bridge can only pass the packets to the bridge. 
* ovs is a software runing in the node
* to make the ovs communicate with the controller, we must leave a veth of the node alone instead of adding it to the ovs bridge.

To make the ovs and ryu connected: 
1. run ovs in a ovs node
2. run controller in another node, called ctrl node
3. add all the veth of the ovs node to the ovs bridge except the veth connecting to the ctrl node.
4. in the ovs node, set the ovs's controller ip&port as the ip of the ctrl node.

for more information, pls ask @JiasenLee

**Reply by PatrickTague** [reply]
awesome, thank you!
